### ISSUE TITLE
F504: Carriage return prevents expression extraction

### ISSUE THREAD

**Issue by addisoncrump** [astral-sh/ruff#4899]
In Python, a standalone carriage return character (0xd or `\r`) is considered whitespace. One could easily argue that this is quite problematic, given that carriage returns can potentially be used to hide parts of source code. That said, it does trigger a failure in F504 (`\r` replaced literally):

```py
"" % {
  'test1': '',\r  'test2': '',
}
```

[And a downloadable version to test against.](https://github.com/charliermarsh/ruff/files/11662694/pathological.txt)

I would simply mark F504 as sometimes fixable, but this solution doesn't feel right. Instead, this feels indicative of incorrect handling of carriage returns as plain whitespace. At the same time, standalone carriage returns should probably be marked as invalid anyways.

I leave this to the discretion of the developers of Ruff :slightly_smiling_face: We should find a fix one way or another to avoid future issues with this being marked as a failure case in the fuzzer.

Discovered by #4822.

**Comment by charliermarsh** [comment]
Perhaps a bug in LibCST, which is failing to parse it?

**Comment by addisoncrump** [comment]
I'm not so clear on that. Is this because libCST is used to parse the expression for repair?

**Comment by charliermarsh** [comment]
Yeah

**Comment by charliermarsh** [comment]
The LibCST parse call throws an error, but it could be our fault, not theirs. We may need to wrap the expression in parentheses before parsing.

**Comment by dhruvmanila** [comment]
> We may need to wrap the expression in parentheses before parsing.

I did a quick experiment and this doesn't help either, it still throws an error:

```rust
libcst_native::parse_expression(&format!("({})", &contents))
```

```
Failed to parse file: Internal error while parsing whitespace: Found newline at (2, 15) but it's not EOL
```

**Comment by MichaReiser** [comment]
This is a bug in LibCST. It seems they don't support mixed newlines:

https://github.com/Instagram/LibCST/blob/c91655fbba7c27db05423673609f10203316b9c0/native/libcst/src/tokenizer/whitespace_parser.rs#L67-L88

**Comment by MichaReiser** [comment]
Upstream fix in LibCST https://github.com/Instagram/LibCST/pull/1007/files

**Comment by MichaReiser** [comment]
The upstream fix is merged. All that we need to do now is to update libcst 🫣
